feat: Profiling Kubernetes Controllers (#85)
mesosphere/d2iq-engineering-blog: content/posts/profiling-kubernetes-controllers-with-pprof/index.md

---

> authors: ["cbuto"]

Reviewer: @cbuto one overall comment is that you're switching 'tense' in this document. Decide if you want to use 'you' or 'we' and then do a search and replace for whichever one you decide. :)

Author: thanks! I updated to use `you`

title: "Profiling Kubernetes Controllers With pprof"

> date: 2023-08-23T16:53:33-04:00

Reviewer 2: Let's try to land this by noon today (SF time)...

```suggestion
date: 2023-08-31T12:00:00-07:00
```

Author: done!

tags: ["kubernetes", "pprof", "kubernetes controllers", "performance"]

> excerpt: Profiling Kubernetes controllers with pprof to help analyze and resolve performance issues

Author: updated to `Analyze and resolve performance issues in Kubernetes controllers using pprof`

> feature_image: feature.png

Author: need to find a better image here

Author: updated

---

# Analyzing Kubernetes controllers performance with pprof

> `pprof` is a Go standard library package which provides tooling for collecting and analyzing profiling data from Go applications. 

Reviewer: ```suggestion
`pprof` is a Go standard library package, which provides tooling for collecting and analyzing profiling data from Go applications. 
```

> Once a profile is collected from an application, it can be analyzed and visualized with the `go tool pprof` command. 

Reviewer: ```suggestion
After a profile is collected from an application, it can be analyzed and visualized with the `go tool pprof` command. 
```

> A common technique for collecting profiles from Go applications is to import the [`"net/http/pprof"`][pprof] 

Author: weird yeah i'll remove the code block!

> package which will register endpoints on an existing HTTP server under the `/debug/pprof/` URL that can be used 

Reviewer: ```suggestion
package, which will register endpoints on an existing HTTP server under the `/debug/pprof/` URL. It can then be used 
```

Author: fixed!

to download live profiles from a running application.

`pprof` can be easily integrated into your Kubernetes controllers to help gain deeper understanding of how a controller
is behaving at runtime with little performance overhead.

### What is a profile? 

> The Godoc for a [`Profile`][pprof profile] describes them as:

Reviewer 2: Codeblocks in links look weird for me; 
![Screenshot 2023-08-31 at 15 18 10](https://github.com/mesosphere/d2iq-engineering-blog/assets/592445/63b0f8b1-72f3-434c-aee1-dc7deefdb782)

to write custom profiles if you have a specific use case that isn't covered in the builtin profiles.
 
The predefined profiles are as follows:
```
goroutine    - stack traces of all current goroutines
heap         - a sampling of memory allocations of live objects
allocs       - a sampling of all past memory allocations
threadcreate - stack traces that led to the creation of new OS threads
block        - stack traces that led to blocking on synchronization primitives
mutex        - stack traces of holders of contended mutexes
```

## Profiling Kubernetes controllers

Now that we know a little bit about `pprof` and profiling, we can look at why we might need this for Kubernetes controllers. Much like
any other application, Kubernetes controllers are prone to suffering from performance issues, running out of memory, etc.

If your controller is being `OOMKilled`, instead of just simply increasing the memory limits and moving on, we can
actually understand what is using up all the memory by collecting and analyzing `heap` or `goroutine` profiles. 

Another example scenario where profiling might help is if a controller is suffering from performance issues when running

> at scale, collecting a `cpu` profile can help identify functions that are using the most CPU time.

Reviewer: ```suggestion
at scale. Collecting a `cpu` profile can help identify functions that are using the most CPU time.
```

Reviewer: other option is a semicolon.

### Enabling `pprof` via controller-runtime

As of `controller-runtime` version v0.15.0, enabling the `pprof` server can be accomplished by specifying the `PprofBindAddress`
option on the controller `manager`. Prior to v0.15.0, it was possible to enable profiling but required manually adding 
each `pprof` endpoint to the existing metrics server via the 
[AddMetricsExtraHandler][AddMetricsExtraHandler] method.

Enabling the `pprof` server on your controller(s) is as simple as this:

```go
opts := ctrl.Options{
    // additional options 
    PprofBindAddress:  "127.0.0.1:8081",
}

mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), opts)
if err != nil {
    setupLog.Error(err, "unable to start manager")
    os.Exit(1)
}
```

I'd recommend always enabling profiling on your Kubernetes controllers by default because you will never know when you need it to debug

> a performance issue until its too late. Keeping it disabled by default will prevent you from to easily debugging performance issues when they pop up because

Reviewer: ```suggestion
a performance issue until its too late. Keeping it disabled by default will prevent you from easily debugging performance issues when they pop up because
```

enabling the `pprof` server will require restarting the pod.

Note: The `pprof` endpoints expose sensitive information so they should always be bound to `127.0.0.1` 

> or kept private by other techniques (i.e. using [kube-rbac-proxy][kube-rbac-proxy])

Reviewer 2: ![Screenshot 2023-08-31 at 19 11 45](https://github.com/mesosphere/d2iq-engineering-blog/assets/592445/f841177b-ced9-48c8-b870-db5d869b6fa2)

> or kept private by other techniques (i.e. using [kube-rbac-proxy][kube-rbac-proxy])

Author: yeah :\ removed them
